### PR TITLE
azurerm_virtual_hub_route_table - add support for default route table

### LINKED ISSUE #10039: azurerm_virtual_hub_route_table needs ability to modify the default route table
The azure resource azurerm_virtual_hub_route_table does not have the ability to modify the default route table. Consider this architecture for reference - https://docs.microsoft.com/en-us/azure/virtual-wan/scenario-route-through-nva

if deploying via terraform, the ability to add routes to the defaultRouteTable is required. However, this resource seems to only allow for CRUD of a new route table.

running this block results in an error message requiring the import of the resource.

resource "azurerm_virtual_hub_route_table" "hub-rt" {
  name           = "defaultRouteTable"
  virtual_hub_id = azurerm_virtual_hub.fgt-vwan-hub.id

  route {
    name              = "route-to-transit"
    destinations_type = "CIDR"
    destinations      = ["10.60.130.0/24"]
    next_hop_type     = "ResourceId"
    next_hop          = azurerm_virtual_hub_connection.transit-vnet-to-hub-connection.id
  }
  route {
    name              = "route-to-spokes"
    destinations_type = "CIDR"
    destinations      = ["10.60.132.0/23","10.60.134.0/23"]
    next_hop_type     = "ResourceId"
    next_hop          = azurerm_virtual_hub_connection.transit-vnet-to-hub-connection.id
  }
}

azurerm_virtual_hub_route_table.hub-rt: Creating...

Error: A resource with the ID "/subscriptions/.../providers/Microsoft.Network/virtualHubs/vwan_hub_fgt/hubRouteTables/defaultRouteTable" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_virtual_hub_route_table" for more information.

  on vwan.tf line 53, in resource "azurerm_virtual_hub_route_table" "hub-rt":
  53: resource "azurerm_virtual_hub_route_table" "hub-rt" {

### PR DESCRIPTION
**Pull request by TomKimber**
Add support for modifying the default route table in azurerm_virtual_hub_route_table resources.

(fixes #10039)

### CODE REVIEW

File: azurerm/internal/services/network/virtual_hub_data_source.go [hunk 1]
```diff
@@ -41,13 +41,65 @@ func dataSourceVirtualHub() *schema.Resource {
 				Computed: true,
 			},
 
+			"default_route_table": {
+				Type:     schema.TypeList,
+				Optional: true,
```
[the remark this answers is not in the document]
Comment (TomKimber) [reply azurerm/internal/services/network/virtual_hub_data_source.go]: Should now be resolved.

File: azurerm/internal/services/network/virtual_hub_resource.go [hunk 2]
```diff
@@ -100,13 +101,78 @@ func resourceVirtualHub() *schema.Resource {
...
+										},
+									},
+
+									"destinations_type": {
+										Type:     schema.TypeString,
+										Required: true,
+										ValidateFunc: validation.StringInSlice([]string{
+											"CIDR",
+											"ResourceId",
+											"Service",
```
[the remark this answers is not in the document]
Comment (TomKimber) [reply azurerm/internal/services/network/virtual_hub_resource.go:146]: The SDK doesn't appear to have constants for destination_type currently - Please see: https://azuresdkdocs.blob.core.windows.net/$web/java/azure-resourcemanager-network/2.2.0/com/azure/resourcemanager/network/models/VirtualHubRouteV2.html#destinationType--

File: azurerm/internal/services/network/virtual_hub_resource.go [hunk 3]
```diff
@@ -100,13 +101,78 @@ func resourceVirtualHub() *schema.Resource {
...
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"labels": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
```
[the remark this answers is not in the document]
Comment (TomKimber) [reply azurerm/internal/services/network/virtual_hub_resource.go:119]: Validation function reference added.

**Comment by TomKimber** [comment]
Fixed Terraform linting - All tests should now pass.

**Comment by TomKimber** [comment]
Ran plugin-sdk-fixerupper.sh to fix merge conflict with master and resolved import issue.

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @TomKimber - looks like we have some test failures:
```
------- Stdout: -------
=== RUN   TestAccVirtualHubConnection_recreateWithSameConnectionName
=== PAUSE TestAccVirtualHubConnection_recreateWithSameConnectionName
=== CONT  TestAccVirtualHubConnection_recreateWithSameConnectionName
    testing.go:620: Step 1/5 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_virtual_hub.test will be updated in-place
          ~ resource "azurerm_virtual_hub" "test" {
                id                  = "/subscriptions/*******/resourceGroups/acctestRG-vhub-210601180507399446/providers/Microsoft.Network/virtualHubs/acctest-VHUB-210601180507399446"
                name                = "acctest-VHUB-210601180507399446"
                # (4 unchanged attributes hidden)
        
              - default_route_table {
                  - labels = [
                      - "default",
                    ] -> null
                }
            }
        
```

**Comment by TomKimber** [comment]
@katbyte This test failure is due to the API requiring the creation of a default route table when the Virtual Hub is created meaning the labels value can't be null. I believe the correct handling of this would be to default the default route table labels set to a set containing "default" which would then create the default route table, would not effect existing users updating the provider who hadn't specified non-default labels and would still allow the removing of the default label if required (Passing an empty set).

It seems that it's not possible currently to default a set or list in the schema (https://github.com/hashicorp/terraform-plugin-sdk/issues/142). How would you advise approaching this?

**Comment by katbyte** [comment]
@TomKimber - in the create function you can detect an empty set and sent that yourself, and if its computed the read should pick it up. but i'd definitly want some good tests to make sure it works

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @TomKimber - running the tests for this we have a bunch of failures:
```
------- Stdout: -------
=== RUN   TestAccVirtualHubConnection_requiresImport
=== PAUSE TestAccVirtualHubConnection_requiresImport
=== CONT  TestAccVirtualHubConnection_requiresImport
    testcase.go:89: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_virtual_hub.test will be updated in-place
          ~ resource "azurerm_virtual_hub" "test" {
                id                  = "/subscriptions/*******/resourceGroups/acctestRG-vhub-210709000609670360/providers/Microsoft.Network/virtualHubs/acctest-VHUB-210709000609670360"
                name                = "acctest-VHUB-210709000609670360"
                # (4 unchanged attributes hidden)
        
              - default_route_table {
                  - labels = [
                      - "default",
                    ] -> null
                }
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccVirtualHubConnection_requiresImport (2393.29s)
FAIL

```

**Comment by manicminer** [comment]
@TomKimber I have rebased your changes and I have some fixes as per katbyte's suggestions - are you able to [allow edits from maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so I can push these and get them tested? Thanks!

**Review by manicminer — COMMENTED** [review]
@TomKimber Additionally, in your `defaultRouteTable` test config there is a circular reference between `azurerm_virtual_hub_connection.test` and `azurerm_virtual_hub.test`. This may require implementation as a separate resource to resolve.

**Comment by timja** [comment]
> @TomKimber I have rebased your changes and I have some fixes as per katbyte's suggestions - are you able to [allow edits from maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so I can push these and get them tested? Thanks!

unfortunately this was submitted from our organisational fork rather than a personal fork.
Organisation forks don't provide that option 😢 

You can send a PR to https://github.com/hmcts/terraform-provider-azurerm/tree/10039-virtual-hub-default-route-table or we can re-submit from someones fork

**Comment by manicminer** [comment]
@timja (cc @TomKimber) Thanks, I'd prefer to close this and open a new PR. Unfortunately however as I mentioned above, in the current state this is unimplementable. This will likely need to be redone as a separate resource, as there is a circular dependency  between `azurerm_virtual_hub` and `azurerm_virtual_hub_connection`:

```hcl
resource "azurerm_virtual_hub" "test" {
  # ...
  default_route_table {
    labels = ["default"]
    route {
      name              = "acctestdefaultroute"
      destinations_type = "CIDR"
      destinations      = ["10.0.2.128/25"]
      next_hop_type     = "ResourceId"
      next_hop          = azurerm_virtual_hub_connection.test.id
    }
  }
}

resource "azurerm_virtual_hub_connection" "test" {
  name                      = "acctestvhubconn"
  virtual_hub_id            = azurerm_virtual_hub.test.id
  remote_virtual_network_id = azurerm_virtual_network.test.id
}
```

**Pull request closed without merging**